### PR TITLE
Fix backend errors handling for fallible syscalls

### PR DESCRIPTION
**Pull request by breathx**
## Error explanation
For example let's take fallible `gr_unreserve_gas` syscall.
https://github.com/gear-tech/gear/blob/9d96bec693b34db229598d0e638502228dc1a43d/core-backend/sandbox/src/funcs.rs#L856-L871
For case if we failed to read read reservation id, atm we are leaving backend with wasm trap, while it should try write ext error about incorrect `reservation_id_ptr` to `write_err_unreserved`, because syscall supposed to be fallible and dance around results.
For case when fails write to `err_<something>` ptr, we already cannot parse error in user space so we should leave with a trap.

## TODO list
- [x] Apply similar changes to wasmi backend
- [x] Provide tests and check that it fails on master

### CODE REVIEW

**Review by ark0f — APPROVED** [review]

File: core-backend/sandbox/src/runtime.rs [hunk 1]
```diff
@@ -91,6 +87,63 @@ where
...
+            *self.err_mut() = err;
+        });
+
+        if body_res.is_err() {
+            if let Ok(to_be_returned) = self.last_err() {
+                body_res = Ok(Err(to_be_returned.encoded_size() as u32));
+            }
+        }
+
+        let body_res = body_res;
```
Comment (grishasobol) [inline core-backend/sandbox/src/runtime.rs:121]: Can you clarify this line please?

**Review by grishasobol — APPROVED** [review]

**Review by grishasobol — CHANGES_REQUESTED** [review]
We make everywhere : 
```rust
ctx.run_fallible(
|ctx| {
    let write_obj = ctx.register_write_as(T)
},
|ctx| {
    // main logic
},
|ctx, write_obj, res| {
    ctx.write_as(write_obj, T::from(res))
},
```

We can make generic method `run_fallible` with T param

File: examples/binaries/backend-error/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+            gsys::gr_send(
+                pid_value.as_ptr(),
+                u32::MAX as *const u8,
+                42,
+                0,
+                res.as_mut_ptr(),
+            )
+        };
+
+        assert!(res.length != 0)
```
Comment (ark0f) [inline examples/binaries/backend-error/src/lib.rs:55]: Also read error to see it's correct

File: gsys/src/lib.rs [hunk 3]
```diff
@@ -170,6 +172,16 @@ impl From<Result<Handle, Length>> for LengthWithHandle {
     }
 }
 
+#[repr(C, packed)]
+#[derive(Default)]
+pub struct LengthBytes([u8; size_of::<Length>()]);
```
Comment (ark0f) [inline gsys/src/lib.rs:177]:
```suggestion
pub struct LengthBytes([u8; mem::size_of::<Length>()]);
```

**Review by ark0f — APPROVED** [review]

**Review by grishasobol — APPROVED** [review]

**Pull request merged**